### PR TITLE
Milestone 10: Authorization with Can Can Can

### PR DESCRIPTION
**Pull request by topeogunleye**
## Milestone 10: Authorization with Can Can Can
In this Pull Request, We:
* Installed CanCanCan in the project.
* Added a role column to the users table.
* Added functionality for user to delete a post if they're logged in and owners of that post. 
  * We added delete button and functionality for comments and posts.

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    # can :destroy, Post, author: user
+    # can :destroy, Comment, author:user
+    # return unless user.admin?
+    # can :manage, :all
+    # Define abilities for the user here. For example:
+    #
+    #   return unless user.present?
+    #   can :read, :all
+    #   return unless user.admin?
+    #   can :manage, :all
```
Comment (arslanbisharat) [inline app/models/ability.rb]: - [x] Let's remove this commented code as it makes code hard to read.

File: app/models/user.rb [hunk 2]
```diff
@@ -3,13 +3,25 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
```
Comment (arslanbisharat) [inline app/models/user.rb:3]: - [x]  Let's remove this commented code as it makes code hard to read.

File: app/models/user.rb [hunk 3]
```diff
@@ -3,13 +3,25 @@ class User < ApplicationRecord
...
-  has_many :likes, foreign_key: :author_id
+  has_many :comments, foreign_key: :author_id, dependent: :delete_all
+  has_many :posts, foreign_key: :author_id, dependent: :delete_all
+  has_many :likes, foreign_key: :author_id, dependent: :delete_all
 
   validates :name, presence: true
   validates :posts_counter, comparison: { greater_than_or_equal_to: 0 }, numericality: { only_integer: true }
 
+  # User::Roles
+  # The available roles
```
Comment (arslanbisharat) [inline app/models/user.rb]: - [x]  Let's remove this commented code as it makes code hard to read.

File: app/models/user.rb [hunk 4]
```diff
@@ -3,13 +3,25 @@ class User < ApplicationRecord
...
+  ROLES = %i[admin default].freeze
+
+  def is?(requested_role)
+    role == requested_role
+  end
+
+  # def admin?
+  #   true
+  # end
```
Comment (arslanbisharat) [inline app/models/user.rb]: - [x]  Let's remove this commented code as it makes code hard to read.

File: db/migrate/20230131230456_add_devise_to_users.rb [hunk 5]
```diff
@@ -39,7 +39,7 @@ def self.up
 
     add_index :users, :email                
     add_index :users, :reset_password_token, unique: true
-    # add_index :users, :confirmation_token,   unique: true
+    add_index :users, :confirmation_token,   unique: true
     # add_index :users, :unlock_token,         unique: true
```
Comment (arslanbisharat) [inline db/migrate/20230131230456_add_devise_to_users.rb:43]: - [x]  Let's remove this commented code as it makes code hard to read.

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
Hi @topeogunleye, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

- The Pull Request has a proper title and description. ✔️
- You have added a descriptive Readme file. Good Job. ✔️
- There is no linter error is present. ✔️
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [ ] Please make sure you right descriptive comments. Commenting involves placing Human Readable Descriptions inside of computer programs detailing what the Code is doing. Proper use of commenting can make code maintenance much easier, as well as helping make finding bugs faster. Further, commenting is very important when writing functions that other people will use. :x:


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Strangeal** [comment]
Thank you @arslanbisharat

**Review by rustam-tolipov — APPROVED** [review]
Hi @topeogunleye, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Developer-512px-10

## Highlights
1. No Linters errors 💯
2. All requirements are complete 💯

## Optional suggestions
No optional suggestions :)

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Strangeal** [comment]
Thank you

**Pull request merged**